app/plugins.go

```diff
@@ -84,3 +88,129 @@ func InitPlugins() {
 		p.OnConfigurationChange()
 	}
 }
+
+func ActivatePlugins() {
+	if Srv.PluginEnv == nil {
+		l4g.Error("plugin env not initialized")
+		return
+	}
+
+	plugins, err := Srv.PluginEnv.Plugins()
+	if err != nil {
+		l4g.Error("failed to start up plugins: " + err.Error())
+		return
+	}
+
+	for _, plugin := range plugins {
+		err := Srv.PluginEnv.ActivatePlugin(plugin.Manifest.Id)
+		if err != nil {
+			l4g.Error(err.Error())
+		}
+		l4g.Info("Activated %v plugin", plugin.Manifest.Id)
+	}
+}
+
+func UnpackAndActivatePlugin(pluginFile io.Reader) (*model.Manifest, *model.AppError) {
+	if Srv.PluginEnv == nil || !*utils.Cfg.PluginSettings.Enable {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
+	}
+
+	filenames, err := utils.ExtractTarGz(pluginFile, Srv.PluginEnv.SearchPath())
```

Comment:
Shouldn't this extract to a subdirectory?

Reply:
I was relying on the tar.gz to have the subdirectory and we need to know the plugin id to create the subdirectory. What I can do is extract to a temporary directory, read the plugin.json, then use the id to move it all to the properly named directory